Fixes problem with deleting all gallery images for products during importing images for products via API POST request.

I didn't found solution for this bug on Magento Github. It is unclear why Magento refuses to delete file(s) from gallery. It is acting like product's image is still used by other store(s), despite of that all of them was deleted from product in each store. In case when images for product(s) are imported by API using POST request to: '/V1/products/:sku/images' endpoint, all images of imported products should be deleted without any checks. This quick fix is solving this problem by trivial, but working way.

This patch is a modification of module-catalog-force-to-delete-all-module-gallery-files-of-product-during-importing-images-for-products-via-api-post-request.patch to provide Magento 2.4.3 compatibility

diff --git a/Model/Product/Gallery/CreateHandler.php b/Model/Product/Gallery/CreateHandler.php
index 7a1bd21..944f087 100644
--- a/Model/Product/Gallery/CreateHandler.php
+++ b/Model/Product/Gallery/CreateHandler.php
@@ -12,6 +12,7 @@ use Magento\Catalog\Model\Product;
 use Magento\Framework\App\Filesystem\DirectoryList;
 use Magento\Framework\App\ObjectManager;
 use Magento\Framework\EntityManager\Operation\ExtensionInterface;
+use Magento\Framework\UrlInterface;
 use Magento\MediaStorage\Model\File\Uploader as FileUploader;
 use Magento\Store\Model\Store;
 use Magento\Store\Model\StoreManagerInterface;
@@ -26,6 +27,8 @@ use Magento\Store\Model\StoreManagerInterface;
  */
 class CreateHandler implements ExtensionInterface
 {
+    public const REQUEST_PATH_WITHOUT_SKU_DURING_PRODUCTS_IMAGES_IMPORT_VIA_API = '/rest/all/V/products//images';
+
     /**
      * @var \Magento\Framework\EntityManager\EntityMetadata
      * @since 101.0.0
@@ -100,6 +103,11 @@ class CreateHandler implements ExtensionInterface
         'thumbnail'
     ];
 
+    /**
+     * @var UrlInterface
+     */
+    protected $url;
+
     /**
      * @param \Magento\Framework\EntityManager\MetadataPool $metadataPool
      * @param \Magento\Catalog\Api\ProductAttributeRepositoryInterface $attributeRepository
@@ -108,6 +116,7 @@ class CreateHandler implements ExtensionInterface
      * @param \Magento\Catalog\Model\Product\Media\Config $mediaConfig
      * @param \Magento\Framework\Filesystem $filesystem
      * @param \Magento\MediaStorage\Helper\File\Storage\Database $fileStorageDb
+     * @param \Magento\Framework\UrlInterface $url
      * @param \Magento\Store\Model\StoreManagerInterface|null $storeManager
      * @throws \Magento\Framework\Exception\FileSystemException
      */
@@ -119,6 +128,7 @@ class CreateHandler implements ExtensionInterface
         \Magento\Catalog\Model\Product\Media\Config $mediaConfig,
         \Magento\Framework\Filesystem $filesystem,
         \Magento\MediaStorage\Helper\File\Storage\Database $fileStorageDb,
+        \Magento\Framework\UrlInterface $url,
         \Magento\Store\Model\StoreManagerInterface $storeManager = null
     ) {
         $this->metadata = $metadataPool->getMetadata(\Magento\Catalog\Api\Data\ProductInterface::class);
@@ -129,6 +139,7 @@ class CreateHandler implements ExtensionInterface
         $this->mediaDirectory = $filesystem->getDirectoryWrite(DirectoryList::MEDIA);
         $this->fileStorageDb = $fileStorageDb;
         $this->storeManager = $storeManager ?: ObjectManager::getInstance()->get(StoreManagerInterface::class);
+        $this->url = $url;
     }
 
     /**
@@ -615,7 +626,9 @@ class CreateHandler implements ExtensionInterface
      */
     private function canRemoveImage(ProductInterface $product, string $imageFile) :bool
     {
-        $canRemoveImage = true;
+        if ($this->isApiImagesImportRequest()) {
+            return true;
+        }
         $gallery = $this->getImagesForAllStores($product);
         $storeId = $product->getStoreId();
         $storeIds = [];
@@ -633,12 +646,32 @@ class CreateHandler implements ExtensionInterface
                     && $image['filepath'] === $imageFile
                     && (int) $image['store_id'] !== $storeId
                 ) {
-                    $canRemoveImage = false;
+                    return false;
                 }
             }
         }
 
-        return $canRemoveImage;
+        return true;
+    }
+
+    /**
+     * It is unclear why Magento refuses to delete file(s) from gallery. It is acting like product's image is still
+     * used by other store(s), despite of that all of them was deleted from product in each store.
+     * In case when images for product(s) are imported by API using POST request to:
+     * /V1/products/:sku/images endpoint, all images of imported products should be deleted without any checks.
+     * This quick fix is solving this problem by trivial, but working way.
+     *
+     * @return bool
+     */
+    protected function isApiImagesImportRequest(): bool
+    {
+        $url = $this->url->getCurrentUrl();
+        $path = parse_url($url, PHP_URL_PATH);
+        $pathWithoutSku = preg_replace('/\d/', '', $path);
+        if ( $pathWithoutSku == self::REQUEST_PATH_WITHOUT_SKU_DURING_PRODUCTS_IMAGES_IMPORT_VIA_API) {
+            return true;
+        }
+        return false;
     }
 
     /**
diff --git a/Model/Product/Gallery/UpdateHandler.php b/Model/Product/Gallery/UpdateHandler.php
index edee9ae..e4169bf 100644
--- a/Model/Product/Gallery/UpdateHandler.php
+++ b/Model/Product/Gallery/UpdateHandler.php
@@ -15,6 +15,7 @@ use Magento\Framework\App\ObjectManager;
 use Magento\Framework\EntityManager\MetadataPool;
 use Magento\Framework\Filesystem;
 use Magento\Framework\Json\Helper\Data;
+use Magento\Framework\UrlInterface;
 use Magento\MediaStorage\Helper\File\Storage\Database;
 use Magento\Store\Model\Store;
 use Magento\Store\Model\StoreManagerInterface;
@@ -41,6 +42,7 @@ class UpdateHandler extends CreateHandler
      * @param Config $mediaConfig
      * @param Filesystem $filesystem
      * @param Database $fileStorageDb
+     * @param UrlInterface $url
      * @param StoreManagerInterface|null $storeManager
      * @param AttributeValue|null $attributeValue
      */
@@ -52,6 +54,7 @@ class UpdateHandler extends CreateHandler
         Config $mediaConfig,
         Filesystem $filesystem,
         Database $fileStorageDb,
+        UrlInterface $url,
         StoreManagerInterface $storeManager = null,
         ?AttributeValue $attributeValue = null
     ) {
@@ -63,6 +66,7 @@ class UpdateHandler extends CreateHandler
             $mediaConfig,
             $filesystem,
             $fileStorageDb,
+            $url,
             $storeManager
         );
         $this->attributeValue = $attributeValue ?: ObjectManager::getInstance()->get(AttributeValue::class);
